File: euler_057.py
from time import time

def sqrt2(n):
    if n == 0:
        return 1
    else:
        return 2


start = time()

A = [1, sqrt2(0)]

B = [0, 1]

for k in range(2,1002):
    A.append(sqrt2(k) * A[k - 1] + A[k - 2])
    B.append(sqrt2(k) * B[k - 1] + B[k - 2])

special = 0

for k in range(2,1002):
    if len(str(A[k])) > len(str(B[k])):
        special += 1

print("Number of fractions: " + str(special))

stop = time()
print("Time: " + str(stop-start))

File: euler_057.js
const start = Date.now();

const numberOfDigits = (bigint) => {
    return bigint.toString().length;
}

let count = 0;

let P = Array(1001);
let Q = Array(1001);

// Initialization and algorithm
// x0 = sqrt(2)
// an = floor(xn), x_(n+1) = 1 / (xn - an)
// P0 = a0, P1 = a0 * a1 + 1, Pn = P_(n-1) * an + P_(n-2)
// Q0 = 1, Q1 = a1, Qn = Q_(n-1) * an + Q_(n-2)

// Simplification: an = 2 for n > 1

P[0] = BigInt(1); P[1] = BigInt(2 * 1 + 1);
Q[0] = BigInt(1); Q[1] = BigInt(2);

for (let n = 2; n < 1001; n++) {
    P[n] = P[n - 1] * BigInt(2) + P[n - 2];
    Q[n] = Q[n - 1] * BigInt(2) + Q[n - 2];

    if (numberOfDigits(P[n]) > numberOfDigits(Q[n])) {
        count++;
    }
}


const end = Date.now();

console.log("Execute time: ", end - start, " Number of fractions is: ", count);